File: linkedListCycle.js
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Step1: Set base cases to not cause the function to fail by an exception
// Step2: Set slow to second value and fast to the third value(slow=1x speed && fast=2x speed)
// Step3: Create the first while loop and have it continue until fast and slow have met
// Step4: After they've met, start the slow pointer back from the beginning or just use head
// to traverse again at 1x speed. We also update fast pointer's speed to become 1x.
// Step5: When the two pointers meet again, that is where the cycle lies.

var detectCycle = function (head) {
  if (head === null || head.next === null || head.next.next === null)
    return null;

  let slowPointer = head.next;
  let fastPointer = head.next.next;

  while (slowPointer !== fastPointer) {
    if (fastPointer.next === null || fastPointer.next.next === null)
      return null;
    slowPointer = slowPointer.next;
    fastPointer = fastPointer.next.next;
  }

  slowPointer = head;
  while (slowPointer !== fastPointer) {
    if (fastPointer.next === null || fastPointer.next.next === null)
      return null;
    slowPointer = slowPointer.next;
    fastPointer = fastPointer.next;
  }

  return fastPointer;
};

File: linked_list_cycle.py
'''
Approach: Take fast and slow pointers to find the intersection point. From head and intersection point start p1 and p2 pointers, and move them to next until p1 equals p2. 

Time: O(n)
Space: O(1)

Passes all testcases in leetcode

'''


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return None
        slow = head
        fast = head
        
        temp=head
        while fast.next and fast.next.next:
            prev = fast
            slow=slow.next
            fast = fast.next.next
            if(slow == fast):
                p1 = head
                p2 = slow
                while(p1!=p2):
                    p1 = p1.next
                    p2 = p2.next
                return p1
                
        else:
            return None
            
        
